File: 24_高级数据结构/good java/problem/FillGridWithL.java
package template.problem;

import template.binary.Bits;
import template.utils.SequenceUtils;

public class FillGridWithL {
    int[][] mat;
    int[][] color;
    int[][] dirs = new int[][]{
            {1, 0},
            {-1, 0},
            {0, 1},
            {0, -1}
    };

    int indicator = 0;
    int[][] ans;

    private void paint1(int i, int j) {
        assert ans[i][j] == 0;
        assert ans[i - 1][j] == 0;
        assert ans[i][j + 1] == 0;
        ans[i][j] = ans[i - 1][j] = ans[i][j + 1] = ++indicator;
    }

    private void paint2(int i, int j) {
        assert ans[i][j] == 0;
        assert ans[i - 1][j] == 0;
        assert ans[i][j - 1] == 0;
        ans[i][j] = ans[i - 1][j] = ans[i][j - 1] = ++indicator;
    }

    private void paint3(int i, int j) {
        assert ans[i][j] == 0;
        assert ans[i + 1][j] == 0;
        assert ans[i][j - 1] == 0;
        ans[i][j] = ans[i + 1][j] = ans[i][j - 1] = ++indicator;
    }

    private void paint4(int i, int j) {
        assert ans[i][j] == 0;
        assert ans[i + 1][j] == 0;
        assert ans[i][j + 1] == 0;
        ans[i][j] = ans[i + 1][j] = ans[i][j + 1] = ++indicator;
    }

    public int[][] solve(int n, int m) {
        indicator = 0;
        boolean flip = false;
        if (n % 3 == 0 && m % 3 != 0) {
            int tmp = m;
            m = n;
            n = tmp;
            flip = !flip;
        }
        if (n % 6 == 0) {
            int tmp = m;
            m = n;
            n = tmp;
            flip = !flip;
        }
        if (n <= 1 || m <= 1 || m % 3 != 0) {
            return null;
        }
        ans = new int[n][m];
        if (n % 2 == 0) {
            for (int i = 0; i < n; i += 2) {
                for (int j = 0; j < m; j += 3) {
                    paint4(i, j);
                    paint2(i + 1, j + 2);
                }
            }
        } else if (m % 2 == 0) {
            for (int i = 0; i < n - 3; i += 2) {
                for (int j = 0; j < m; j += 3) {
                    paint4(i, j);
                    paint2(i + 1, j + 2);
                }
            }
            for (int i = 0; i < m; i += 2) {
                paint4(n - 3, i);
                paint2(n - 1, i + 1);
            }
        } else if (m >= 9 && n >= 5) {
            paint4(0, 0);
            paint4(2, 0);
            paint2(4, 1);
            paint3(1, 2);
            paint3(0, 3);
            paint1(4, 2);
            for (int i = 3; i + 2 <= m - 2; i += 2) {
                paint4(2, i);
                paint1(4, i + 1);
            }
            paint3(3, m - 1);
            paint2(2, m - 1);
            paint4(0, m - 2);
            for (int i = 4; i < m - 2; i += 3) {
                paint1(1, i);
                paint3(0, i + 2);
            }
            for (int i = 5; i < n; i += 2) {
                for (int j = 0; j < m; j += 3) {
                    paint4(i, j);
                    paint2(i + 1, j + 2);
                }
            }
        } else {
            return null;
        }
        if (flip) {
            ans = transpose(ans);
        }
        return ans;
    }

    private int[][] transpose(int[][] mat) {
        int n = mat.length;
        int m = mat[0].length;
        int[][] ans = new int[m][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                ans[j][i] = mat[i][j];
            }
        }
        return ans;
    }

    private void setColor(int i, int j, int v, int fi, int fj, int c) {
        if (v != mat[i][j]) {
            return;
        }
        color[i][j] = c;
        for (int[] d : dirs) {
            int x = d[0] + i;
            int y = d[1] + j;
            if (!valid(x, y)) {
                continue;
            }
            if (x == fi && y == fj) {
                continue;
            }
            setColor(x, y, v, i, j, c);
        }
    }

    private boolean valid(int i, int j) {
        return i >= 0 && j >= 0 && i < mat.length && j < mat[0].length;
    }

    private int findColor(int i, int j, int v, int fi, int fj) {
        if (v != mat[i][j]) {
            if (color[i][j] == -1) {
                return 0;
            }
            return 1 << color[i][j];
        }
        int ans = 0;
        for (int[] d : dirs) {
            int x = d[0] + i;
            int y = d[1] + j;
            if (!valid(x, y)) {
                continue;
            }
            if (x == fi && y == fj) {
                continue;
            }
            ans |= findColor(x, y, v, i, j);
        }
        return ans;
    }

    private int[][] color(int[][] mat) {
        int n = mat.length;
        int m = mat[0].length;
        this.mat = mat;
        color = new int[n][m];
        SequenceUtils.deepFill(color, -1);
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (color[i][j] == -1) {
                    int bits = findColor(i, j, mat[i][j], -1, -1);
                    int c = 0;
                    while (Bits.get(bits, c) == 1) {
                        c++;
                    }
                    setColor(i, j, mat[i][j], -1, -1, c);
                }
            }
        }
        return color;
    }
}

File: 11_动态规划/经典题/相邻行间dp/轮廓线dp/FillGridWithL.go
package main

import "fmt"

func main() {
	G := NewFillGridWithL()
	res := G.Solve(5, 6)
	fmt.Println(res)
	color := G.Color(res)
	fmt.Println(color)
}

var DIR4 = [][]int32{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}

// 蒙德里安的梦想：用L形块填充格子.
// 多米诺平铺.
type FillGridWithL struct {
	res       [][]int32
	mat       [][]int32
	color     [][]int32
	indicator int32
}

func NewFillGridWithL() *FillGridWithL {
	return &FillGridWithL{}
}

func (f *FillGridWithL) Solve(n, m int32) [][]int32 {
	f.indicator = 0
	flip := false
	if n%3 == 0 && m%3 != 0 {
		n, m = m, n
		flip = !flip
	}
	if n%6 == 0 {
		n, m = m, n
		flip = !flip
	}
	if n <= 1 || m <= 1 || m%3 != 0 {
		return nil
	}
	f.res = make([][]int32, n)
	for i := int32(0); i < n; i++ {
		f.res[i] = make([]int32, m)
	}
	if n%2 == 0 {
		for i := int32(0); i < n; i += 2 {
			for j := int32(0); j < m; j += 3 {
				f.paint4(i, j)
				f.paint2(i+1, j+2)
			}
		}
	} else if m%2 == 0 {
		for i := int32(0); i < n-3; i += 2 {
			for j := int32(0); j < m; j += 3 {
				f.paint4(i, j)
				f.paint2(i+1, j+2)
			}
		}
		for i := int32(0); i < m; i += 2 {
			f.paint4(n-3, i)
			f.paint2(n-1, i+1)
		}
	} else if m >= 9 && n >= 5 {
		f.paint4(0, 0)
		f.paint4(2, 0)
		f.paint2(4, 1)
		f.paint3(1, 2)
		f.paint3(0, 3)
		f.paint1(4, 2)
		for i := int32(3); i+2 <= m-2; i += 2 {
			f.paint4(2, i)
			f.paint1(4, i+1)
		}
		f.paint3(3, m-1)
		f.paint2(2, m-1)
		f.paint4(0, m-2)
		for i := int32(4); i < m-2; i += 3 {
			f.paint1(1, i)
			f.paint3(0, i+2)
		}
		for i := int32(5); i < n; i += 2 {
			for j := int32(0); j < m; j += 3 {
				f.paint4(i, j)
				f.paint2(i+1, j+2)
			}
		}
	} else {
		return nil
	}
	if flip {
		f.res = f.transpose(f.res)
	}
	return f.res
}

func (f *FillGridWithL) Color(mat [][]int32) [][]int32 {
	n, m := int32(len(mat)), int32(len(mat[0]))
	f.mat = mat
	f.color = make([][]int32, n)
	for i := int32(0); i < n; i++ {
		f.color[i] = make([]int32, m)
		for j := int32(0); j < m; j++ {
			f.color[i][j] = -1
		}
	}
	for i := int32(0); i < n; i++ {
		for j := int32(0); j < m; j++ {
			if f.color[i][j] == -1 {
				bits := f.findColor(i, j, mat[i][j], -1, -1)
				c := int32(0)
				for bits>>c&1 == 1 {
					c++
				}
				f.setColor(i, j, mat[i][j], -1, -1, c)
			}
		}
	}
	return f.color
}

func (f *FillGridWithL) paint1(i, j int32) {
	f.indicator++
	f.res[i][j] = f.indicator
	f.res[i-1][j] = f.indicator
	f.res[i][j+1] = f.indicator
}

func (f *FillGridWithL) paint2(i, j int32) {
	f.indicator++
	f.res[i][j] = f.indicator
	f.res[i-1][j] = f.indicator
	f.res[i][j-1] = f.indicator
}

func (f *FillGridWithL) paint3(i, j int32) {
	f.indicator++
	f.res[i][j] = f.indicator
	f.res[i+1][j] = f.indicator
	f.res[i][j-1] = f.indicator
}

func (f *FillGridWithL) paint4(i, j int32) {
	f.indicator++
	f.res[i][j] = f.indicator
	f.res[i+1][j] = f.indicator
	f.res[i][j+1] = f.indicator
}

func (f *FillGridWithL) transpose(mat [][]int32) [][]int32 {
	n, m := int32(len(mat)), int32(len(mat[0]))
	res := make([][]int32, m)
	for i := int32(0); i < m; i++ {
		res[i] = make([]int32, n)
	}
	for i := int32(0); i < n; i++ {
		for j := int32(0); j < m; j++ {
			res[j][i] = mat[i][j]
		}
	}
	return res
}

func (f *FillGridWithL) valid(i, j int32) bool {
	return i >= 0 && j >= 0 && i < int32(len(f.mat)) && j < int32(len(f.mat[0]))
}

func (f *FillGridWithL) setColor(i, j, v, fi, fj, c int32) {
	if v != f.mat[i][j] {
		return
	}
	f.color[i][j] = c
	for _, d := range DIR4 {
		x, y := d[0]+i, d[1]+j
		if !f.valid(x, y) {
			continue
		}
		if x == fi && y == fj {
			continue
		}
		f.setColor(x, y, v, i, j, c)
	}
}

func (f *FillGridWithL) findColor(i, j, v, fi, fj int32) int32 {
	if v != f.mat[i][j] {
		if f.color[i][j] == -1 {
			return 0
		}
		return 1 << f.color[i][j]
	}
	res := int32(0)
	for _, d := range DIR4 {
		x, y := d[0]+i, d[1]+j
		if !f.valid(x, y) {
			continue
		}
		if x == fi && y == fj {
			continue
		}
		res |= f.findColor(x, y, v, i, j)
	}
	return res
}
